proposals/mdm-baseline.md

```diff
@@ -0,0 +1,153 @@
+# MDM core
+
+This document is a high-level spec/guide of the technical challenges that don't
+directly map to a product feature request.
+
+Some of this might already be captured in other issues, but we wanted to put
+everything together and have multiple people thinking about the solutions as a
+whole.
+
+We will create issues to tackle actionable bits, but the idea is to also use
+this to guide the implementation of feature requests.
+
+
+### Problem: Hosts can be enrolled via MDM
+
+Currently, there's no reliable way to associate an MDM enrolled device with a
+host in the `hosts` table.
+
+- If a host is enrolled in MDM before its enrolled via `osquery` (eg: DEP
+  enrollment) we will have a record in the `nano_enrollments` table before the
+  host exists in the `hosts` table.
+
+- If a host exists in both tables, we could `JOIN` them using `hosts.uuid =
+  nano_enrollments.device_id`  but we need to make sure the right constraints
+  in the application logic exist and the right indexes are created for fast
+  access.
+
+Currently, the code thinks of these two tables as two disjoint sets of elements
+that represent the same entity: a host.
+
+Motivation:
+
+- In https://github.com/fleetdm/fleet/issues/8878 we're asked to return hosts
+  that are scheduled for DEP enrollment (not even enrolled yet) to the `GET
+  /hosts` response.
+- In https://github.com/fleetdm/fleet/issues/8878 we're asked to retrieve a
+  list of hosts enrolled via `osquery` that are pending to be enrolled via MDM.
+
+#### Proposed solution
+
+`hosts` should always be the source of truth, we could consider
+`nano_enrollments` the same way we treat  the "auxiliary" `hosts_*` tables (eg:
+`host_disks`): as tables holding metadata for a host. 
+
+When we get information through MDM channels (be it via an enrollment, or via
+the DEP API notifying a new device purchased,) we always get a device serial
+number.
+
+1. Ensure there's a host in the `hosts` table with that serial number (insert
+   one if there isn't.) Try to pre-populate as much info as possible from the
+   MDM payload (eg: `ProductName` is `hosts.hardware_model`)
+1. Create a new entry in the related `nano_*` table pointing to the host.
+
+This implies that we must add an uniqueness constraint for the serial number in both tables.
+
+We should also add a new column or table to track the MDM enrolling status
+("pending," "enrolled") without having to `JOIN` both tables.
+
+#### TODO:
+
+1. Investigate what are the non-optional fields in the `hosts` table, and if
+   having default empty values for some of them requires a breaking change.
+1. Investigate the consequences of using a serial number to uniquely identify
+   hosts, eg: what happens if two users on the same machine use different
+   enrollment profiles? See "macOS extensions" in MDM spec for details.
```

Comment:
Well... thinking about this a bit more, if we DON'T add the constraint in the hosts table, we should only have 1 unenrolled host from DEP with that serial number (because it's new hardware... how would it have been duplicated?) so what if we just grab the first one we get from the hosts table and log if there are more.

Reply:
I'm OK with this too 👍